### DISCUSSION TITLE
Idea: Change the behavior of the up and down arrow keys without losing any functionality with many benefits.

### DISCUSSION THREAD

**Discussion by aaa1386** [freeplane/freeplane#2169] [Ideas]
<gray>Watch the video below. Nodes with green icons will move in the same column when using the up and down arrow keys, no matter how many times they are used, and this behavior is recommended.</gray>

https://github.com/user-attachments/assets/2d38fdd2-a5b8-4d54-9520-0cea7ec578b2

<gray>In nodes with red cross icons, when we move with the down arrow key, we are moved to another column at the end. If we use the up arrow key, we do not return to the initial node that had the red icon. This behavior may confuse the user. Pay attention to the three starting points of navigation with the cross icon. Is it easy for you to imagine how this works or confusing?</gray>

<gray>Imagine that you want to navigate between brothers and sisters in a large map where a node has many children. What happens if you hold down the down arrow key to quickly reach the end of the column? If you do not pay attention to the lowest node, you may get lost and be transported to another part of the map. There is no way back because the up arrow key does not return you to the original path. This problem has happened to me many times and I have to be careful not to go past the wrong bottom node.</gray>

<gray>Now, create a large map and quickly move by pressing the down button. In this case, you don't really know where you are on the map and every time you press the down button, you get more lost. Now, suppose you want to return to your starting point with the up button. This is practically impossible and makes you even more confused. In addition to vertical movement, the up and down arrow keys also move left and right, which makes no sense and confuses the user. It would be better if the arrow keys only moved vertically and did not move to other columns.</gray>

<b>Suggestion for solving the problem:</b> 
In Freeplane, there are two options called `Navigate → Select Up` and `Navigate → Select Down` that perform the current function of scrolling up and down. But the up and down arrow keys on the keyboard work independently of these two options. That is, even if no shortcut is defined for these two options, the arrow keys will still have their current function of scrolling up and down.

<gray>It is recommended that the arrow keys be defined so that they only scroll through siblings and not exit them, similar to the function you saw in the video for the three nodes with the green icon.</gray>

<gray>If some users do not want to lose the current functionality of Freeplane, there is no problem. They can define their own shortcuts for these two options (`Navigate → Select Up` and `Navigate → Select Down` ). They can even set the up and down arrow keys for them. Therefore, by implementing this idea, no functionality of Freeplane will be reduced.</gray>

<b>Benefits of the new idea:</b>

<b>You can move to the end or beginning of the sibling column at full speed.</b>

<gray>More focus:</gray> 
The user is more focused when navigating a branch because they no longer have to worry about being unexpectedly taken to other parts of the map.

<b>Maintain ease and speed in navigation:</b> 
With this change, the user can quickly reach the beginning or end of a branch without worrying about making a mistake while navigating between columns. This feature is especially useful in large and complex maps.

<b>Simple and clear:</b> 
By limiting the movement of the directional buttons to the same column, navigation will be clearer and simpler for the user.

<b>Ability to return to the starting point:</b> 
The user will always be able to return to their starting point using the directional buttons and avoid getting lost in the map.

<b>More flexibility for users:</b> 
Users who want to move between columns can set their own shortcuts for the two options `Navigate → Select Up` and `Navigate → Select Down`  to maintain the current functionality of the FP. As a result, nothing is removed from the current functionality of FreePlane.

<b>Reduced Confusion:</b> 
This change can help prevent user confusion and provide a better user experience, especially when the map is large and complex.

<b>Conclusion:</b> 
Implementing this change in FreePlane will make navigating between nodes easier and error-free. These changes also give the user more control over how the map is navigated, without requiring major changes to the current functionality of the software. This feature will improve the user experience, increase navigation speed and accuracy, and ultimately increase the efficiency of mind maps.

**Comment by dpolivaev** [comment]
There is a dedicated action to go back to the previously selected node, Navigate->Go backward (Alt+←)

**Comment by aaa1386** [comment]
> There is a dedicated action to go back to the previously selected node, Navigate->Go backward (Alt+←)

Yes, that's right - but the functionality that was proposed is really very simple and interesting.
Look at the image below
Columns 1, 2 and 3 are known.

![image](https://github.com/user-attachments/assets/86d6ceb0-0241-4ab7-a7de-29a841f3f1e4)


 Suppose the user navigates in column 2 with the up and down arrow keys. Whenever he decides to move to column 1 or 3, he moves with the right and left arrow keys and navigates again with the up and down arrow keys. That is, there is always a control mode that does not unconsciously navigate to unwanted nodes and does not get confused

Can't the up and down arrow keys or the two options that exist for navigating and selecting the up and down nodes ( `Navigate → Select Up` and `Navigate → Select Down` ) be changed to suit the idea, and the other one remains for the current state of the FP?

**Comment by aaa1386** [comment]
In my opinion, it is even better to use the `Shift + Up` and `Shift + Down` button combination to select nodes only among siblings. Whenever the user decides to select from other branches, by holding the Control button and clicking on another branch, he is moved to it, and there he can again select other nodes using the `Shift + Up` and `Shift + Down` buttons.

The current functionality of FP in using the `Shift + Up` and `Shift + Down` keys to select multiple nodes often causes errors for the user. You can try it, you should always be careful not to make any mistakes.

**Comment by aaa1386** [comment]
Current functionality: When you navigate with the down arrow key, if you reach the end of the siblings, for example node 888, you will be moved to the highest node of the uncle of these siblings, which is below their father.
The up arrow key is the opposite of this functionality.

Question: How often is such a confusing functionality used? How useful is it?
I have never used this function, either in navigation by `up` and `down` keys  or in selecting a number of nodes with `Shift + Up` and `Shift + Down` , but I was always careful not to let it lead me into error.
Others, I don't know if it really has a specific use for it or not?

**Comment by aaa1386** [comment]
Look at the video below: Consider nodes 8, 7, and 10


https://github.com/user-attachments/assets/eb63aa7b-7850-469c-9b2a-60ae4d0685a6




We navigate to node 8 with the down arrow key
We press the key again
If node 7 is unfolded, node 10 is selected
If node 7 is folded, node 7 is selected
It seems a bit confusing.

I always have to look at nodes like node 9 first to see if its uncle (i.e. node 7) is folded or unfolded so that I can figure out which node is selected by pressing the down arrow key.

Guess other states like the up key function

**Reply by dpolivaev** [reply]
If you press an arrow key e.g. "down", it goes to the closest visible node below the currently selected node.
I do not see what could be easier.

**Comment by dpolivaev** [comment]
I converted the issue to discussion because I think other people can easier find it here and participate.

**Comment by aaa1386** [comment]
Suppose I want to select from node `1-->3` to the last node of the siblings. Using the` Shift + Down` key, I get to the end of the siblings. I accidentally press Shift + Down again. See what happens. Node 7 is selected. I have no way to go back. If I press `Shift + Up` , it gets worse because node 1 is also selected, making it harder to go back.
I have encountered this problem many times.

https://github.com/user-attachments/assets/91ece207-62f3-47dc-afda-13e76e46efbc

**Reply by dpolivaev** [reply]
Selection modification using shift is different from just navigation. In this case I like your idea to stay inside siblings better.
On the other side, what should happen if node 7 were unfolded?

**Reply by aaa1386** [reply]
> Selection modification using shift is different from just navigation. In this case I like your idea to stay inside siblings better. On the other side, what should happen if node 7 were unfolded?

I don't understand what you mean, but in any case, the ` Shift+Up` or  `Shift+ Down` button should not go outside the sibling range.
 If the user wants to open a node from another column in addition to nodes `1-->3` to node `9` , he must activate the other column with the Control button and click. Then, if he wants, he can use the `Shift+Up` or  `Shift+ Down`  button.

 Suppose you have three paragraphs in MicrosoftWord. You want to select only the first and third paragraphs. What do you do? You can use  `Shift+ Down`  for the first paragraph. Then, you must activate it by Control-clicking on a part of the third paragraph. Then, you can use `Shift+ Down`  again.

**Comment by euu2021** [comment]
About the navigation: If I understood correctly, this script creates the described behavior:

For the down arrow:

```
if(node.root) return

nodePosition = node.parent.getChildPosition(node)
nextSibling = node.parent.children[nodePosition + 1]


if (nextSibling == null || node == c.viewRoot) {
    menuUtils.executeMenuItems([ 'FreeScrollAction.UP' ])
}

else {
    c.select(nextSibling)
}
```

For the up arrow:

```
if(node.root) return

nodePosition = node.parent.getChildPosition(node)
previousSibling = node.parent.children[nodePosition - 1]

if (nodePosition == 0 ||  node == c.viewRoot) {
    menuUtils.executeMenuItems([ 'FreeScrollAction.DOWN' ])
}

else {
    c.select(previousSibling)
}
```

If it's correct, I can create the script for the selection, too.

**Reply by aaa1386** [reply]
That's right - thanks - if possible, write for multiple node selection mode.

**Reply by euu2021** [reply]
I updated my comment, to improve the scripts. Now, it deals with the root, and root views.

> That's right - thanks - if possible, write for multiple node selection mode.

Ok

**Comment by euu2021** [comment]
Script for the selection:

For the shift+down arrow:

```
import org.freeplane.features.mode.Controller

if(node.root) return

nodePosition = node.parent.getChildPosition(node)
nextSibling = node.parent.children[nodePosition + 1]


if (nextSibling == null || node == c.viewRoot) {
    menuUtils.executeMenuItems([ 'FreeScrollAction.UP' ])
}

else {
    Controller.currentController.mapViewManager.mapView.getMapSelection().selectContinuous(nextSibling.delegate)
}
```

For the shift+up arrow:

```
import org.freeplane.features.mode.Controller

if(node.root) return

nodePosition = node.parent.getChildPosition(node)
previousSibling = node.parent.children[nodePosition - 1]


if (nodePosition == 0  || node == c.viewRoot) {
    menuUtils.executeMenuItems([ 'FreeScrollAction.DOWN' ])
}

else {
    Controller.currentController.mapViewManager.mapView.getMapSelection().selectContinuous(previousSibling.delegate)
}
```

**Comment by dpolivaev** [comment]
@euu2021 I see that you can craft a script helping Alij, and I still do not know how you find the discussed idea itself. Should the logic of selection be changed in generally and when it is used with the shift?

Interestingly, I have got a related request https://github.com/freeplane/freeplane/issues/2142 suggesting a different way of change: using arrow up after arrow down should always bring the selection back instead of going to the nearest upper node as now. Actually, I removed this logic as I introduced new node layouts, but I think I know how I could implement it. But I am not sure if it would simplify things and which idea should be taken.

As you know, I do not use Freeplane often enough to have my own opinion.

Any thoughts?

**Reply by euu2021** [reply]
> Should the logic of selection be changed in generally and when it is used with the shift?

I've been testing this new setup today, and I like it. As it is easier to understand, I can press the navigation buttons really fast, because I know where the selection will stop. This way, I end up navigating the map. The current navigation behavior is not intuitive at all, for me.

For the selection, the new rubber band selection covers that function of selecting across branches. So, I think the behavior can be changed for the selection, too.

**Comment by aaa1386** [comment]
See  #2174

**Comment by quickfold** [comment]
@dpolivaev I agree with the other posters that the suggested behavior works better than the current behavior. You wrote, "If you press an arrow key e.g. "down", it goes to the closest visible node below the currently selected node. I do not see what could be easier." You are right that this behavior is easy to understand, but it is not so easy to use when navigating large maps. In a map where a parent has a lot of children, often when you use the down arrow and go below the lowest child, and it selects an uncle node, and you by reflex press the up arrow, but that takes you somewhere new, not where you want to go. Often I am pressing the down arrow 10+ times and it's easy to press it 2-3 extra times by accident and I get totally lost as to where the currently selected node is in relation to the end of the sibling nodes you were trying to select. 

You are right that I can use a shortcut to find the last selected node, and that's what I currently do, but for me it requires a mental shift from "navigate in directional space" to "navigate in time back to a previous time".

I think it would be easier to navigate by keyboard if the up and down arrows had 'stops' built in so the user could more easily navigate by keyboard without getting lost. Up and down would move up and down among siblings only, and left and right would move up and down a generation. This is an easier mental model for me for navigation by keyboard than the current behavior. Thanks for considering.

**Comment by dpolivaev** [comment]
I  changed the logic as suggested in https://github.com/freeplane/freeplane/issues/2142. I think that this solution satisfies your needs.

Preview 1.12.9_01

**Reply by aaa1386** [reply]
> I changed the logic as suggested in #2142. I think that this solution satisfies your needs.
> 
> Preview 1.12.9_01


Yes, now if you just use the up button, you can return to the starting point.  It is good change but ...


I will name this comment "Eliminating Spiral Movements" in navigation using the up and down buttons.

We are discussing the functionality of the up button; the down operation works the same but in reverse.  

Look at the image below and focus on level 3. 

![image](https://github.com/user-attachments/assets/dd4abf7d-2377-4411-a6cc-2070f900b9ed)

The starting point is the lowest node of level 3 (blue color). This lowest node has a parent and uncles, all located in level 2 (yellow color). Level 2 represents general keywords with their respective nodes beneath them. Typically, nodes at level 2 share relationships and similarities, which is why they are grouped at this level.

**Current Functionality:**  
When navigating upwards from the lowest point of level 3 using the up button, upon reaching the highest point of the sibling list:  
- If the uncle node is folded, the navigation shifts to the uncle's branch (promoting it to level 2).  
- If the uncle node is unfolded, navigation continues upward in the same third column.  

This functionality is not user-friendly. When a user at level 2 has folded some nodes and unfolded others, it means they are focusing only on the unfolded nodes and their children in the third column. Why should a node in the second column be selected by the up button during navigation at the third level, especially when the user is ignoring it?  

**Proposed Suggestion:**  
I propose a new functionality with clear logic. We hope it will be considered. If anyone has a simpler and clearer suggestion, please share it.  

The up or down button movement should be restricted to a single level and never leave that level, regardless of whether the uncles are folded or unfolded.  

**Advantages (refer to the image):**  
1. Nodes folded at level 2 (yellow) are not selected because the user is ignoring them.  

2. Movement using the up or down button always follows a straight and clear line within the same level.  

3. Returning to the starting point is easy, and the user has a clear understanding of this return. For instance, if they navigate upward in level 3, they can move back down in the same level, with all navigated connectors clearly visible.  

4. A specific level (e.g., level 3) can be easily selected using the **Shift** and **Up** buttons. If a node in the second column is folded, its children are not selected, as the user has intentionally folded it to ignore them. So you can easily select the obvious nodes in a level. I think this is the easiest way to select a level. With the current state of FP, what are the options for choosing a long level?

5. ...

**Reply by aaa1386** [reply]
If you want the whole idea to be complete, create two options for the user to choose from:
    1. Navigation with the up and down buttons limited to siblings only.
    2. Navigation with the up and down buttons across the entire level.

**Comment by euu2021** [comment]
I like the navigation limited for siblings, for the reasons that I explained at https://github.com/freeplane/freeplane/discussions/2169#discussioncomment-11671206

And, I will add to that the idea of making a cycle between the siblings. I.e., if the bottom sibling is selected, and the down arrow is pressed, the top sibling is selected. This would allow quick navigation through long lists of siblings. Also, this would help to solve [this difficult situation](https://github.com/freeplane/freeplane/issues/2174#issuecomment-2564675114) described by ajaberi1360, because the selection could continue in the top sibling.

Anyway, I think it's complicated to understand the new behavior, that goes from fff to hhh in the example below:

![image](https://github.com/user-attachments/assets/f5f39fb8-86d3-4e25-9684-7ea0536794bf)

Those jumps between uncle and nephews are too surprising to the mind. If the user went from eee to fff, and, some hours later, he goes back to the map and presses down arrow, the selection will jump to hhh, and the user will get confused.

**Reply by aaa1386** [reply]
> 
> And, I will add to that the idea of making a cycle between the siblings. I.e., 

One of the nice benefits of cycle behavior is that if you reach the end of a long list of nodes, you can easily return to the beginning of the list by pressing the Down button once. I encounter this problem quite often. Currently, I have to navigate a large number of nodes to reach the beginning of the list.

**Reply by aaa1386** [reply]
>  Also, this would help to solve [this difficult situation](https://github.com/freeplane/freeplane/issues/2174#issuecomment-2564675114) 
> 
Yes, it's interesting that it can solve a problem.

**Comment by quickfold** [comment]
Is there a way to jump to the top or bottom of a list of siblings? I often have nodes with many siblings and want to go to the top or bottom without worrying about overshooting. If there's a way to do that automatically, I can, but if there's not, then I actually would prefer if the behavior was not to cycle, but just to stop when pressing up at the top or down at the bottom.

**Reply by aaa1386** [reply]
> Is there a way to jump to the top or bottom of a list of siblings? I often have nodes with many siblings and want to go to the top or bottom without worrying about overshooting. If there's a way to do that automatically, I can, but if there's not, then I actually would prefer if the behavior was not to cycle, but just to stop when pressing up at the top or down at the bottom.

Yes, there is a way.
See the problem I showed in the video below when the user moves from the parent node (1-->1) to the child node with the RIGHT button,  moved to the node that was last selected among the siblings (1-->1-->2). Not to the first node of the siblings (1-->1-->1).
This function is useful in some cases, for example, when the user has recently navigated that branch, but if they have navigated that branch a long time ago, this function will confuse the user.
This function creates another problem that is mentioned in issue #2174.
It is suggested to create a function that allows the user to navigate to the first node in the siblings  list by pressing the left or right button. (At least the user can choose whether to execute the old function or this new function) If such a function is created, it will always cause you to be moved to the beginning of the list. To go to the end of the list, you can use the cycle behavior, i.e. press the up button once. Which takes you to the end of the list.


https://github.com/user-attachments/assets/0d0bfa03-efcf-445a-9f5c-60fc3732965f

**Reply by aaa1386** [reply]
Of course I guess FP can't tell which node is the first in the list, yes we know which one is the first in the list because it is obvious to us, but for FP any node that has already been selected among the siblings will be considered the beginning of the list. I think
FP can never change its habit.
So the only possible way to access the beginning and end of a list more easily is to implement this idea:
Scrolling up and down buttons only among siblings, also having a cycle behavior

**Reply by quickfold** [reply]
If I understand you, there is currently NOT a way to quickly move to the first or last sibling in a long list, so you are proposing that someone add that function to Freeplane. I would also like this function added to Freeplane. However, I would still want to change the current behavior, as it is so easy to lose your place in a map when the selection moves to an uncle node. This happens to me several times a day and is very frustrating.

We are discussing how to balance 1) the ability to easily select the first or last sibling without overshooting (stop at top or bottom, don't cycle) with 2) how to move from the top of the list to the bottom (or vice versa) with the fewest keystrokes (don't stop at top or bottom, but cycle). If someone adds a select top/bottom sibling function, then option 2) is no longer needed, and FP should not cycle.

In any relatively short list (i.e. visible on one screen), you can move to top or bottom with a single press of page up/page down. This means that neither stop at top/bottom nor cycle are needed to easily move to the top or bottom of a group of siblings. However, with a big group of siblings where you can't see both top and bottom siblings at the same time, currently you need multiple keystrokes to navigate to top/bottom. So with multiple keystrokes needed, the change of accidentally pressing too many times is quite likely.

Currently, if you are trying to get to the top/bottom quickly with page up/down and you accidentally press page up/down a 2nd time, the selection moves to an uncle, and it is harder to navigate back to where you want to be. If you move up/down with arrow keys, and there is a cycle behavior, you are likely to overshoot and navigate past the top or bottom and need to navigate back. It will be a bit confusing since you will jump past where you want to be, but it is easier to navigate back than with page up/down because it wouldn't move you to an uncle.

The ideal solutions from a user perspective for solving our keyboard navigation problem is to change the current behavior from moving from the top/bottom sibling to an uncle node, and instead offer a behavior setting for "when navigating selection above top sibling or below bottom sibling" and letting the user choose "stop" or "cycle". 

I don't know how hard that would be to implement. Perhaps @dpolivaev can tell us if he is interested in adding this setting, and if so, how hard it would be?

**Reply by dpolivaev** [reply]
I have already added an option to stop at the first or last sibling to the source code, which will be available in the next preview version.

Could you please help me better understand the situations where jumping to the first or last sibling is important for you? Additionally, I’d like to understand why you work with so many siblings that they don’t fit on one screen and why accidentally pressing the key multiple times becomes a significant issue for you.

I often struggle to imagine how you use the program since you don’t share enough details about your workflow. To improve the program and address your needs, I need to fully understand your perspective, which is only possible if you explain it in detail. Your input is essential for making the program more effective for your use.

**Reply by euu2021** [reply]
> The ideal solutions from a user perspective for solving our keyboard navigation problem is to change the current behavior from moving from the top/bottom sibling to an uncle node, and instead offer a behavior setting for "when navigating selection above top sibling or below bottom sibling" and letting the user choose "stop" or "cycle".

A compromise may be: 
Scenario: selection is at the last sibling.
- if the user presses down once: nothing happens. Ie, no cycle behavior.
- if the user presses down twice, in quick succession, the selection cycles (goes to the top sibling).

So, to open the portal and teleport to the first sibling, the user needs to press down twice. The same works when the selection is on the top sibling.

**Reply by dpolivaev** [reply]
Preview 1.12.9_03

**Reply by quickfold** [reply]
Hi @dpolivaev ,

> I have already added an option to stop at the first or last sibling to the source code, which will be available in the next preview version.

Great, thanks.
 
> Could you please help me better understand the situations where jumping to the first or last sibling is important for you? Additionally, I’d like to understand why you work with so many siblings that they don’t fit on one screen and why accidentally pressing the key multiple times becomes a significant issue for you.

Jumping to first or last provides fast navigation in long lists, which is useful. However, my request for that function was more because it would help me prevent overshooting the first or last item when trying to get to the top or bottom of a list quickly. Overshooting these nodes currently moves me to an uncle node, which happens to me a lot (maybe 5+ times a day), and is disorienting and frustrating, because I have to figure out the parent node of the last node I was trying to select.

As far as the list not fitting on one screen. One of my main uses of Freeplane is for long-form writing (e.g. 20k+ words). Sometimes my nodes are tall because they include paragraphs of writing, and if I have several under one parent, I can't see all of them. Just FYI, I have a 27" 4k 16:9 screen, so the problem isn't that I'm on a small laptop screen.

I also use FP for planning, and I might want a list of 40 items relating to meetings over 6 months, and for each I might need to see 1-3 nodes of description in the list. That looks like the screenshot below. Perhaps this image also helps you imagine why letting the user control spacing between nodes, as discussed in the "compact map" idea a few months ago, could be useful. Let me know if you have other questions.

![image](https://github.com/user-attachments/assets/32314e88-fe01-4519-a0fd-c811ef7377f0)

**Reply by quickfold** [reply]
I'm glad to see the new behavior in 1.2.9 pre4! I had trouble finding the new setting due to the label, "sibling selection by keyboard". Also, the label "changes parent" is potentially confusing because it sounds like the selected node will change parent. It is hard to describe what this setting does in just a few words and in a way users will understand. I will propose the following revisions:

Name of setting: replace "Sibling selection by keyboard" with "In a list of sibling nodes, when the top/bottom node is selected, what should up/down arrow do?"

I'm not sure what the current selection behavior's logic is, but I propose replacing "Changes parent" with either "Select nearest visible non-sibling node" or "Select sibling of selection's parent node" (whichever is correct).

I'd suggest replacing "loops over siblings" with "Loop selection to bottom/top sibling", and replacing "Stays at the end" with "Nothing, keep current selection".

**Reply by aaa1386** [reply]
> Preview 1.12.9_03

The behavior of the "Up" and "Down" buttons during navigation has been updated, offering three options for users.   Thanks.

When the "Shift" key is pressed and the "Up" and "Down" buttons are used simultaneously, multiple nodes can be selected.  
It seems that the behavior of the "Up" and "Down" buttons is similar whether they are used alone or in combination with the "Shift" key. Therefore, there should be a clear and logical relationship between these functionalities.  

However, despite my efforts, I could not find a clear logic for the behavior of the "Up" and "Down" buttons when used with the "Shift" key. In the attached video, I’ve demonstrated an example of incorrect behavior.  

https://github.com/user-attachments/assets/2e3ebd27-2ebd-45d7-83ad-5617758cc8a5


I suggest that the functionality set by the user for the Up and Down buttons should also apply when selecting multiple nodes using the Shift key in combination with the Up or Down buttons.

**Reply by dpolivaev** [reply]
I do not know how to reproduce it. I have never seen Freeplane working as in the video.

**Reply by aaa1386** [reply]
> I do not know how to reproduce it. I have never seen Freeplane working as in the video.

Follow the steps below
1- Select node 1
2- Select node 1-1 by pressing the Shift key and right-clicking
3- Press the Down key while holding the Shift key
4- Press the Down key again while holding the Shift key
See the incorrect behavior

[TEST.zip](https://github.com/user-attachments/files/18382001/TEST.zip)

**Reply by aaa1386** [reply]
Watch the video below (at the end of the video)
Does the shift and left button work correctly? It should not perform the function of the shift and up button.

https://github.com/user-attachments/assets/e1772d90-460b-4be4-ba80-4a2e8841a7e1

**Reply by aaa1386** [reply]
There are three modes for the Up and Down navigation buttons. Whichever mode the user selects, the following controls must remain consistent and adhere to the same pattern:

Page Up and Page Down buttons
Page Up and Page Down buttons combined with the Shift key
Up and Down buttons combined with the Shift key

https://github.com/user-attachments/assets/8ead1460-55ba-4f3a-802b-bd8f9b7dec7f

**Reply by aaa1386** [reply]
Watch the video bleow.  
Assume that the "Loop over Siblings" mode is selected:  
1. **Select the node AAAAAAAAAA.**  
2. **Press Shift and Page Up.** In this case, the nodes above the node AAAAAAAAAA will be selected.  
3. **Press Shift and Page Down.** Ideally, the selected nodes should be deselected, but this does not happen.  
4. **If, instead of step 3 (pressing Shift and Page Down), you press Shift and Page Up,** all sibling nodes should be selected (since the "Loop over Siblings" mode is active). However, the FP software does not perform this action.  


https://github.com/user-attachments/assets/9ee5fbf5-ba5c-4929-bbbd-7eb3ca7c06e8

**Reply by aaa1386** [reply]
I posted some videos .
I think the idea proposed in the link https://github.com/freeplane/freeplane/discussions/2169#discussioncomment-11802411 is a very good proposed solution and makes the FP function unified and logical.

**Comment by quickfold** [comment]
Request: can the new behavior setting "Sibling selection by keyboard" apply not just to up/down arrows, but also page up/page down buttons? To move quickly through lists, I usually use page up/down, but these still have the old selection behavior when used at the top/bottom of lists.

**Reply by dpolivaev** [reply]
Done

**Comment by aaa1386** [comment]
Is it possible to fix the problem I mentioned in the two links below? I sent you a map in the same link to  test the problem.

https://github.com/freeplane/freeplane/discussions/2169#discussioncomment-11802219
https://github.com/freeplane/freeplane/discussions/2169#discussioncomment-11801041

**Comment by aaa1386** [comment]
Many navigation issues have been resolved with recent updates.  
Previously, I suggested that the behavior of the **Page Up** and **Page Down** keys, when used with the **Shift** key, should match their behavior when used without the **Shift** key.  
This means that if the user selects the option `Sibling navigation via keyboard --> Stop at boundary siblings`, the behavior of the **Page Up** and **Page Down** keys, whether used with or without the **Shift** key, should be the same.  

However, I have now realized that using the **Shift** key together with the **Page Up** or **Page Down** keys in the `Sibling navigation via keyboard --> Stop at boundary siblings` mode is neither useful nor logical.  
The reason is that when the **Shift** key is combined with the **Page Up** or **Page Down** keys, the user is in multi-node selection mode, and stopping at the end of the list conflicts with the user's goal in this mode.  

You can test this yourself by enabling this mode and using the **Shift + Page Up/Page Down** keys with multiple nodes. You'll see what I mean.  I think no one likes the behavior of the `Sibling navigation via keyboard → Stop at boundary siblings` option when using Shift along with keyboard keys like arrow keys, Page Up, and Page Down.

Therefore, I propose that when the user selects the `Sibling navigation via keyboard --> Stop at boundary siblings` option, the behavior of the **Shift + Page Up/Page Down** keys should follow the logic of the `Cycle through siblings` option. This won't confuse users because they naturally expect to achieve two modes when using **Shift + Page Up/Page Down**:  
 `Cycle through siblings`   OR
 `Navigate across all same-level siblings`  

Adding a third behavior, such as `Sibling navigation via keyboard --> Stop at boundary siblings`, has no practical use for the user. This is because when the **Shift + Page Up/Page Down** keys are used, the user is in multi-selection mode, and stopping at the end of the list contradicts their goal. However, if a cycling behavior is present, it works perfectly without any issues.  

This adjustment would simply make the behavior of **Shift + Page Up/Page Down** more logical for this specific scenario.  


### Second Proposal:  

Extend the same behavior adjustment I proposed for the **Shift + Page Up/Page Down** keys to also apply to the **Shift + Up/Down Arrow** keys. This would create a more logical and consistent behavior for managing and selecting nodes across the same level or siblings when using **Shift + Up/Down Arrow** keys.  


 Users can easily select multiple nodes from the same level or sibling nodes.By using the **Shift** key in combination with the **Up** and **Down** arrow keys, users can efficiently navigate and select nodes. These two modes provide flexibility, allowing users to choose between navigating through all nodes at the same level or cycling within sibling nodes as needed.  

Users can perform this action by switching between two distinct selection modes:  
1. `Navigate across all same-level siblings`  
2. `Cycle through siblings`

**Reply by aaa1386** [reply]
> Many navigation issues have been resolved with recent updates. Previously, I suggested that....

When the user holds the Shift key to select multiple nodes, they should have maximum flexibility in choosing nodes. Restricting the selection to only among siblings ، not only limits the user but also doesn’t provide any significant benefit. 
It seems that with the help of the Shift key and the Page Up and Page Down keys, the user can easily achieve their desired states.  

A more logical solution might be to configure the functionality of the Shift key along with the Up, Down, Page Up, and Page Down keys to enable navigation across all same-level siblings. This would provide maximum flexibility for the user when selecting multiple nodes without losing any useful functionality.  

Based on the tests I’ve conducted on the features introduced in the recent updates, it seems that the approach I’ve suggested in this comment would be more effective and logical.

In fact, this comment and the previous comment suggest two new scenarios for improving performance. Even if the suggestion that corresponds to this comment is not implemented, it is necessary to implement the suggestions in the previous comment.

_**I prefer implementing this comment over the previous one and think it is more logical and useful.**_
